contracts/staking-vault/StakingVault.sol

```diff
@@ -0,0 +1,266 @@
+  }
+
+  function addRewards(address tokenBacked, uint256 rewardAmount) external {
+    _addRewards(rewardAmount, tokenBacked);
+    IERC20(REWARD_TOKEN).safeTransferFrom(msg.sender, address(this), rewardAmount);
+
+    emit AddRewards(tokenBacked, rewardAmount);
+  }
+
+  function accumulateRewards(address tokenBacked, uint256 rewardAmount) external {
+    undistributedRewards[tokenBacked] += rewardAmount;
+    IERC20(REWARD_TOKEN).safeTransferFrom(msg.sender, address(this), rewardAmount);
+
+    emit AccumulateRewards(tokenBacked, rewardAmount);
+  }
+
+  function distributeRewards(address tokenBacked) external {
+    uint256 rewards = undistributedRewards[tokenBacked];
+    _addRewards(rewards, tokenBacked);
+    undistributedRewards[tokenBacked] = 0;
+
+    emit DistributeRewards(tokenBacked, rewards);
+  }
+
+  function claimRewards(uint256[] memory tokenIDs) external {
+    for (uint256 i; i < tokenIDs.length; ) {
+      _claimReward(tokenIDs[i]);
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  function _updatePrice(address tokenBacked, uint256[] memory amountsDrawn) internal {
+    address[] memory tokensStaked = _stakingTokens;
+    uint256 previousBalance;
+    for (uint256 i = 0; i < tokensStaked.length; ) {
+      uint256 tokenID = convertToID(tokensStaked[i], tokenBacked);
+      previousBalance = balanceStakedPerID[tokenID];
+      balanceStakedPerID[tokenID] = previousBalance - amountsDrawn[i];
+      _sTokenPrice[tokenID] = (_sTokenPrice[tokenID] * (previousBalance - balanceStakedPerID[tokenID])) / previousBalance;
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  function _amountToMint(uint256 tokenID, uint256 amount) internal view returns (uint256) {
+    uint8 decimals = tokenDecimalCache[identifierToTokens[tokenID].depositToken];
+
+    return ((amount * 1e18) / _sTokenPrice[tokenID]) * 10**(18 - decimals);
+  }
+
+  function _amountToSend(uint256 tokenID, uint256 amountBurnt) internal view returns (uint256) {
+    uint8 decimals = IERC20Metadata(identifierToTokens[tokenID].depositToken).decimals();
+
+    return ((_sTokenPrice[tokenID] * amountBurnt) / 1e18) / 10**(18 - decimals);
+  }
+
+  function _addRewards(uint256 rewardAmount, address tokenBacked) internal {
+    address[] memory tokensStaked = _stakingTokens;
+    uint256 totalValue;
+    uint256[] memory values = new uint256[](tokensStaked.length);
+    uint256 tokenID;
+    for (uint256 i; i < tokensStaked.length; ) {
+      tokenID = convertToID(tokensStaked[i], tokenBacked);
+      values[i] = PRICE_FEED.queryReturn(tokensStaked[i], VALUATION_TOKEN, balanceStakedPerID[tokenID]);
+      totalValue += values[i];
+      unchecked {
+        ++i;
+      }
+    }
+    for (uint256 i; i < tokensStaked.length; ) {
+      tokenID = convertToID(tokensStaked[i], tokenBacked);
+      values[i] = (values[i] * 1e18) / totalValue;
+      rewardsPerID[tokenID] += (rewardAmount * values[i]) / supplyPerID[tokenID];
```

Comment:
I wonder if rewardAmount is USDC. 1e6 won't that break the logic? do we need to do some decimal calcs here?

Reply:
no it wont because the decimals for percentage of rewardAmount cancel out and it will be left to be based in whatever decimal value rewardAmount is.